sky/provision/instance_setup.py

```diff
@@ -0,0 +1,347 @@
+    log_path_abs = str(provision_config.config.provision_log)
+    logger.info(f'Running command on head node: {_MAYBE_SKYLET_RESTART_CMD}')
+    returncode, stdout, stderr = ssh_runner.run(_MAYBE_SKYLET_RESTART_CMD,
+                                                stream_logs=False,
+                                                require_outputs=True,
+                                                log_path=log_path_abs)
+    if returncode:
+        raise RuntimeError('Failed to start skylet on the head node '
+                           f'(exit code {returncode}). Error: '
+                           f'===== stdout ===== \n{stdout}\n'
+                           f'===== stderr ====={stderr}')
+
+
+@_auto_retry
+def _internal_file_mounts(file_mounts: Dict,
+                          runner: command_runner.SSHCommandRunner,
+                          log_path: str) -> None:
+    if file_mounts is None or not file_mounts:
+        return
+
+    for dst, src in file_mounts.items():
+        # TODO: We should use this trick to speed up file mounting:
+        # https://stackoverflow.com/questions/1636889/how-can-i-configure-rsync-to-create-target-directory-on-remote-server
+        full_src = os.path.abspath(os.path.expanduser(src))
+
+        if os.path.isfile(full_src):
+            mkdir_command = f'mkdir -p {os.path.dirname(dst)}'
+        else:
+            mkdir_command = f'mkdir -p {dst}'
+
+        rc, stdout, stderr = runner.run(mkdir_command,
+                                        log_path=log_path,
+                                        stream_logs=False,
+                                        require_outputs=True)
+        subprocess_utils.handle_returncode(
+            rc,
+            mkdir_command, ('Failed to run command before rsync '
+                            f'{src} -> {dst}.'),
+            stderr=stdout + stderr)
+
+        runner.rsync(
+            source=src,
+            target=dst,
+            up=True,
+            log_path=log_path,
+            stream_logs=False,
+        )
+
+
+@_log_start_end
+def internal_file_mounts(cluster_name: str, common_file_mounts: Dict,
+                         cluster_metadata: common.ClusterMetadata,
+                         ssh_credentials: Dict[str,
+                                               str], wheel_hash: str) -> None:
+    """Executes file mounts - rsyncing internal local files"""
+    _hint_worker_log_path(cluster_name, cluster_metadata,
+                          'internal_file_mounts')
+
+    def _setup_node(runner: command_runner.SSHCommandRunner,
+                    metadata: common.InstanceMetadata, log_path: str):
+        del metadata
+        _internal_file_mounts(common_file_mounts, runner, log_path)
+
+    _parallel_ssh_with_cache(_setup_node,
+                             cluster_name,
+                             stage_name='internal_file_mounts',
+                             digest=wheel_hash,
```

Comment:
wheel_hash is not enough, as we may upload some changes of the credential files. The following commands triggers some error in the `provision.log`:
```
sky launch -c test-skypilot-install --cloud aws --cpus 2
sky launch -c test-skypilot-install
```

THe `provision.log` has the following error:

```
I 09-03 16:54:56 instance_setup.py:71] --------------------Start: internal_file_mounts --------------------
D 09-03 16:54:56 metadata_utils.py:64] Need to run stage internal_file_mounts on instance i-0d27d2f2b5f398c93 of cluster test-skypilot-install-9ce1: False
I 09-03 16:54:56 instance_setup.py:75] --------------------End:   internal_file_mounts --------------------
I 09-03 16:54:56 instance_setup.py:75] 
I 09-03 16:54:56 instance_setup.py:71] 
I 09-03 16:54:56 instance_setup.py:71] --------------------Start: run_runtime_setup_on_cluster --------------------
D 09-03 16:54:56 metadata_utils.py:64] Need to run stage run_runtime_setup_on_cluster on instance i-0d27d2f2b5f398c93 of cluster test-skypilot-install-9ce1: True
cp: cannot stat '/home/ubuntu/.sky/.runtime_files/81806728-5be9-4f33-9611-b6a27d791b71': No such file or directory
```

Reply:
do you mean combine it with `wheel_hash`?

Reply:
The wheel path is already included in the `file_mounts` dict right?